File: frontend/src/app/personal-data.ts
export interface PersonalData {
    id: string;
    firstname: string;
    surname: string;
    email: string;
    address?: string;
    place?: string;
    city?: string;
    province?: string;
    notes?: string;
}

/** A dictionary containing the mapping from a key of PersonalData to a displayable string */
export const personalDataKeysToDisplayMap: { [P in keyof PersonalData]-?: string } = {
    "id": "Id",
    "firstname": "First name",
    "surname": "Surname",
    "email": "Email",
    "address": "Address",
    "place": "Place",
    "city": "City",
    "province": "Province",
    "notes": "Notes",
};

export function isPersonalDataKey(key: string): key is keyof PersonalData {
    return !!Object.keys(personalDataKeysToDisplayMap).find(k => k == key);
}

File: backend/src/main/java/org/acme/models/PersonalData.java
package org.acme.models;

import java.sql.ResultSet;
import java.sql.SQLException;

public record PersonalData(Long id, String firstname, String surname, String email, String address, String place,
                           String city, String province, String notes) {

    @Override
    public String firstname() {
        return Character.toUpperCase(firstname.charAt(0)) + firstname.substring(1).toLowerCase();
    }

    @Override
    public String surname() {
        return Character.toUpperCase(surname.charAt(0)) + surname.substring(1).toLowerCase();
    }

    public boolean isValid() {
        return firstname != null
                && !firstname().isBlank()
                && !firstname.contains(" ")
                && surname != null
                && !surname().isBlank()
                && !surname.contains(" ")
                && email != null
                && email().matches("^[\\w-\\.]+@([\\w-]+\\.)+[\\w-]{2,4}$");
    }

    public static PersonalData fromResultSet(final ResultSet resultSet) throws SQLException {
        return new PersonalData(
                resultSet.getLong("id"),
                resultSet.getString("firstname"),
                resultSet.getString("surname"),
                resultSet.getString("email"),
                resultSet.getString("address"),
                resultSet.getString("place"),
                resultSet.getString("city"),
                resultSet.getString("province"),
                resultSet.getString("notes")
        );
    }
}
